app/src/main/java/com/eis/phoneringer/MainActivity.java

```diff
@@ -1,14 +1,127 @@
+        } else {
+            requestPermission();
+        }
+
+        /**
+         * Default password equals for all the devices
+         */
+        if (passwordManager.getPassword().isEmpty())
+            passwordManager.setPassword(DEFAULT_PASSWORD);
+
+        ringButton.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+                sendCommand(context);
+            }
+        });
+    }
+
+    /**
+     * Sends a ring command to the phone number given by the user
+     *
+     * @param context The current application context
+     */
+    private void sendCommand(final Context context) {
+        //Callback when the command is sent
+        SMSSentListener smsSentListener = new SMSSentListener() {
+            @Override
+            public void onSMSSent(SMSMessage message, SMSMessage.SentState sentState) {
+                Toast.makeText(context, "Command sent to " + phoneNumber.getText().toString(), Toast.LENGTH_SHORT).show();
+            }
+        };
+        //Passing to the AppManager a new RingCommand built through the phone number given by the user
+        try {
+            AppManager.getInstance().sendCommand(context, new RingCommand(new SMSPeer(phoneNumber.getText().toString()), passwordManager.getPassword()), smsSentListener);
+        } catch (InvalidTelephoneNumberException e) {
+            Toast.makeText(context, "Invalid phone number", Toast.LENGTH_SHORT).show();
+        } catch (InvalidSMSMessageException e) {
+            Toast.makeText(context, "Invalid SMSMessage object", Toast.LENGTH_SHORT).show();
+        }
     }
+
+    /**
+     * @return true if the app has permissions, false otherwise
+     */
+    public boolean checkPermission() {
+        if (!(getApplicationContext().checkSelfPermission(Manifest.permission.RECEIVE_SMS) == PackageManager.PERMISSION_GRANTED) ||
+                !(getApplicationContext().checkSelfPermission(Manifest.permission.RECEIVE_SMS) == PackageManager.PERMISSION_GRANTED))
+            return false;
+        else
+            return true;
+    }
+
+    /**
+     * Checks if permissions are granted, if not then requests them to the user
+     */
+    public void requestPermission() {
+        if (!checkPermission())
+            requestPermissions(new String[]{Manifest.permission.SEND_SMS, Manifest.permission.RECEIVE_SMS}, 0);
+    }
+
+    @Override
+    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
+        if (grantResults[0] == PackageManager.PERMISSION_GRANTED && grantResults[1] == PackageManager.PERMISSION_GRANTED) {
+            SMSHandler.getInstance().setup(getApplicationContext());
+            SMSHandler.getInstance().setReceivedListener(new ReceivedMessageListener(getApplicationContext()));
+        } else {
+            finish();
+            System.exit(0);
```

Comment:
what do these two lines do? do they terminate execution of the app? `finish()` terminates the current Activity, while I have no idea of what `System.exit(0)` does in Android, in Java it terminates execution of the current JVM.
why are you terminating the current Activity?

Reply:
Well I don't know sincerely.
Time ago I saw using this method of "exit" by @CremaLuca and since then I have been using it

Reply:
My fault, I read the wrong answer on this https://stackoverflow.com/questions/6330200/how-to-quit-android-application-programmatically
Should have used `finishAndRemoveTask()` since we're using API 21+.

Reply:
What I would do in that `else` block would be display a Toast telling the user that we need those permissions in order for the app to work, and then call `requestPermission()` again

Reply:
I'll try your solution @giovannivelludo